Add support for password protected key files to ssl_server2 and ssl_client2

hanno-arm opened the pull request:
__Summary:__ This PR adds support for password protected key files in the `ssl_server2` and `ssl_client2` example applications through the additional command line parameters `key_pwd`. It also makes a cosmetic change of debug output on the client and server along the way.

This came up while attempting to reproduce https://github.com/ARMmbed/mbedtls/issues/1411.

k-stachowiak commented on ChangeLog:
```diff
@@ -23,6 +23,8 @@ Features
      may be up to two bytes shorter. This allows the library to support all
      hash and signature sizes that comply with FIPS 186-4, including SHA-512
      with a 1024-bit key.
+   * Add key_pwd command line parameter to ssl_server2 and ssl_client2
```
Negligible remark: two arguments are added to `ssl_server`, so maybe this should be formulated more generally, without refering to `key_pwd`, and hence omitting `key_pwd2`, or more precisely, taking all the new parameters into account.

hanno-arm replied:
I'll reformulate it.

hanno-arm replied:
Fixed.

k-stachowiak commented on ChangeLog:
```diff
@@ -23,6 +23,8 @@ Features
      hash and signature sizes that comply with FIPS 186-4, including SHA-512
      with a 1024-bit key.
+   * Add key_pwd command line parameter to ssl_server2 and ssl_client2
+     example application to support the use of password-protected key files.
```
"application" should probably be "applications".

hanno-arm replied:
Yes!

hanno-arm replied:
Fixed.

k-stachowiak commented on library/ssl_cli.c:
```diff
@@ -918,8 +918,8 @@ static int ssl_write_client_hello( mbedtls_ssl_context *ssl )
 
-        MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, add ciphersuite: %04x",
-                                    ciphersuites[i] ) );
+        MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, add ciphersuite: %04x (%s)",
```
Since we print hexadecimal values, maybe it would be better to prefix them with `0x`, so that we don't lead to ambiguity, when e.g. "42" is printed.

hanno-arm replied:
Yes, that's worth fixing along the way, too :+1:

hanno-arm replied:
Fixed.

hanno-arm commented:
@k-stachowiak Thank you for your review! I believe to have addressed your comments. Could you please re-review?

@mpg Could you please re-review, too?

k-stachowiak commented on library/ssl_srv.c:
```diff
@@ -817,7 +817,8 @@ static int ssl_ciphersuite_match( mbedtls_ssl_context *ssl, int suite_id,
     }
 
-    MBEDTLS_SSL_DEBUG_MSG( 3, ( "trying ciphersuite: %s", suite_info->name ) );
+    MBEDTLS_SSL_DEBUG_MSG( 3, ( "trying ciphersuite: %04x (%s)",
```
I'm sorry not to have annotated this before. The change here doesn't account for the fact of printing hexadecimal numbers. @hanno-arm could you please fix this as well? Or maybe I am missing a reason for the code being this way here?

hanno-arm replied:
Fixed.

hanno-arm commented:
@k-stachowiak Please re-review.

hanno-arm commented:
@k-stachowiak I rebased this on top of `development`, resolving conflicts in `ssl_client2.c`. Could you please re-review?

@AndrzejKurek Do you a few minutes to review this, too?

hanno-arm commented:
@AndrzejKurek @k-stachowiak Thank you for your review! I've rebased this on top of `development`. Could you please re-review?

AndrzejKurek commented on ChangeLog:
```diff
@@ -103,6 +103,12 @@ Changes
      suggests). #2671
    * Make `make clean` clean all programs always. Fixes #1862.
+   * Add the command line parameter key_pwd to the ssl_client2 and ssl_server2
+     example applications which allows to provide a password for the key file specified
```
Lines too long.

hanno-arm replied:
Fixed.

danh-arm commented:
Hi @hanno-arm . I've just spotted this stalled, almost complete PR. Is there any chance you can rebase this (using the new Changelog entry format) so we can push this through?

hanno-arm commented:
@danh-arm Sure, I'll rebase.

hanno-arm commented:
@danh-arm Done.

@AndrzejKurek @mpg Would you mind reviewing this again?

danh-arm commented on ChangeLog:
```diff
@@ -46,6 +46,13 @@ Changes
+     these applications with password-protected key files. Analogously but for
+     ssl_server2 only, add the command line parameter key_pwd2 which allows to
+     set a password for the key file provided through the existing key_file2
+     argument.
```
We've moved to using ChangeLog entries instead of directly editing the ChangeLog to avoid conflicts. More guidance here: https://github.com/ARMmbed/mbedtls/blob/development/ChangeLog.d/00README.md

danh-arm commented:
Sorry, there are some CI failures, e.g. :
```
  Gen   test_suite_cipher.ccm.c

:1927:25: error: string literal of length 4215 exceeds maximum length 4095 that ISO C99 compilers are required to support [-Werror,-Woverlength-strings]

        mbedtls_printf( USAGE2 );

                        ^~~~~~

ssl/ssl_server2.c:497:5: note: expanded from macro 'USAGE2'

    "    auth_mode=%%s        default: (library default: none)\n"      \

    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
and 
```
ECJPAKE: server not configured ......................................... FAIL

  ! pattern 'add ciphersuite: c0ff' MUST be present in the Client output

  ! outputs saved to o-XXX-420.log

ECJPAKE: working, TLS .................................................. FAIL

  ! pattern 'add ciphersuite: c0ff' MUST be present in the Client output

  ! outputs saved to o-XXX-421.log
```

danh-arm commented:
Hi @hanno-arm. Will you have time to pick up this old PR again?

hanno-arm commented:
Hi @danh-arm, Yes, I'll pick it up again.

AndrzejKurek commented:
I'll also be ready for review once the current issues are resolved.

hanno-arm commented:
Great, thanks @AndrzejKurek !

mpg commented:
This is also on my review list, so we should be able progress this quickly :)

hanno-arm commented:
@mpg @AndrzejKurek I rebased the PR and adapted the ChangeLog format. No other changes were made. Could you please re-review?

hanno-arm commented:
@mpg @AndrzejKurek I pushed a commit addressing that I hadn't adjusted the expected debug strings in `ssl-opt.sh` after changing the way ciphersuites are reported in the debug log. 

Could you please take another look (once the CI is [hopefully] happy)?

mpg commented:
The CI now seems happy about the `ssl-opt.sh` messages, but complains about `USAGE2` growing to long - this is because we now enforce standard compliance even in the example programs.

hanno-arm commented:
@mpg I noticed that the PR leaves a test-gap, because the new command line options are never exercised. I think it's preferable to add those tests before merging the PR - WDYT?

mpg commented:
@hanno-arm Good point, thanks for noticing - I should have thought about it while reviewing. I think it should be easy enough to add a test in `ssl-opt.sh` exercising the new options, and it would be better to have that before merging the PR indeed.

hanno-arm commented:
@mpg @AndrzejKurek Done. Could you please re-review?
